src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -560,7 +568,50 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                     }
 
                     break;
+                }
+                case 3:
+                {
+                    assert(instrIsRMW);
+                    assert(targetReg != falseReg);
+                    assert(targetReg != embMaskOp2Reg);
+                    assert(targetReg != embMaskOp3Reg);
+                    assert(!HWIntrinsicInfo::IsOptionalEmbeddedMaskedOperation(intrinEmbMask.id));
+
+                    if (intrin.op3->IsVectorZero())
+                    {
+                        // If `falseReg` is zero, then move the first operand of `intrinEmbMask` in the
+                        // destination using /Z.
+
+                        assert(targetReg != embMaskOp2Reg);
+                        GetEmitter()->emitIns_R_R_R(INS_sve_movprfx, emitSize, targetReg, maskReg, embMaskOp1Reg, opt);
 
+                        // Finally, perform the actual "predicated" operation so that `targetReg` is the first operand
+                        // `embMaskOp2Reg` is the second operand and `embMaskOp3Reg` is the third operand.
+                        GetEmitter()->emitIns_R_R_R_R(insEmbMask, emitSize, targetReg, maskReg, embMaskOp2Reg,
+                                                      embMaskOp3Reg, opt);
+                    }
+                    else
+                    {
+                        // If the instruction just has "predicated" version, then move the "embMaskOp1Reg"
+                        // into targetReg. Next, do the predicated operation on the targetReg and last,
+                        // use "sel" to select the active lanes based on mask, and set inactive lanes
+                        // to falseReg.
+
+                        assert(HWIntrinsicInfo::IsEmbeddedMaskedOperation(intrinEmbMask.id));
+
+                        if (targetReg != embMaskOp1Reg)
+                        {
+                            GetEmitter()->emitIns_R_R(INS_sve_movprfx, EA_SCALABLE, targetReg, embMaskOp1Reg);
+                        }
+
+                        GetEmitter()->emitIns_R_R_R_R(insEmbMask, emitSize, targetReg, maskReg, embMaskOp2Reg,
+                                                      embMaskOp3Reg, opt);
+
+                        GetEmitter()->emitIns_R_R_R_R(INS_sve_sel, emitSize, targetReg, maskReg, targetReg, falseReg,
+                                                      opt, INS_SCALABLE_OPTS_UNPREDICATED);
```

Comment:
Is there an assumption being made about the instruction being RMW here?

`FMLA` encodes 4 registers (`Zda`, `Pg`, `Zn`, and `Zm`) where `Zda` is both the source and destination and the operation is functionally similar to `Zda += (Zn * Zm)` (with only a single rounding operation).

Given some `Zda = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), Zda)` it can then be encoded as simply:
```asm
fmla Zda, Pg/M, Zn, Zm
```

Given some `Zda = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), merge)` it can then be encoded as simply:
```asm
movprfx Zda, Pg/M, merge
fmla Zda, Pg/M, Zn, Zm
```

Given some `Zda = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), Zero)` it can then be encoded as simply:
```asm
movprfx Zda, Pg/Z, Zda
fmla Zda, Pg/M, Zn, Zm
```

There are then similar versions possible using `fmad` when the multiplier is the source and destination (`op2Reg == tgtReg` or `op3Reg == tgtReg`).

-----------------

We should actually never need `sel` for this case, but only need complex generation if `tgtReg` is unique from all input registers (including the merge) and we're merging with a non-zero value, such as `dest = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), merge)`:
```asm
mov dest, Zda
movprfx dest, Pg/M, merge
fmla dest, Pg/M, Zn, Zm
```

This ends up being different from the other fallbacks that do use `sel` specifically because it's `RMW` and requires predication (that is there is no `fmla (unpredicated)`).

Reply:
The main goal of using `ins (unpredicated); sel` in the other case is because it allows a 2 instruction sequence as the worst case.

In this case, we at worst need a 3 instruction sequence due to the required predication on the instruction. Thus, it becomes better to use `mov; movprfx (predicated); ins (predicated)` instead as it can allow `mov` to be elided by the register renamer.

Reply:
>  such as dest = ConditionalSelect(Pg, FusedMultiplyAdd(Zda, Zn, Zm), merge):

For the similar reasoning mentioned in https://github.com/dotnet/runtime/pull/100743#discussion_r1577289650 (where we should only `movprfx` the inactive lanes from `merge -> dest`, the code should be:


```asm
mov dest, Zda
fmla dest, Pg/M, Zn, Zm
sel dest, Pg/M, dest, merge

```

Reply:
Actually, I misinterpreted the value of `Pg/M` as AllTrue. Spoke to @tannergooding offline and we would like to generate:

```asm
sel dest, Pg/M, Zda, merge
fmla dest, Pg/M, Zn, Zm
```